ARROW-3610: [C++] Add interface to turn stl_allocator into arrow::MemoryPool

xhochy opened the pull request:
First outcome from PyConDE collaboration with @wolfv

pitrou commented on cpp/src/arrow/allocator.h:
```diff
@@ -85,6 +85,41 @@ class stl_allocator {
+  ~STLMemoryPool() override = default;
+
+  Status Allocate(int64_t size, uint8_t** out) override {
+    *out = alloc_.allocate(size);
```
Shouldn't we try to catch `std::bad_alloc` here and return `OutOfMemory`?

pitrou commented on cpp/src/arrow/allocator.h:
```diff
@@ -85,6 +85,41 @@ class stl_allocator {
+    uint8_t* old_ptr = *ptr;
+    *ptr = alloc_.allocate(new_size);
+    memcpy(*ptr, old_ptr, std::min(old_size, new_size));
+    Free(old_ptr, old_size);
```
This will wrongly call `UpdateAllocatedBytes`.

pitrou replied:
Let's say that before reallocate, you have `bytes_allocated = old_size`.
After reallocate, you want to have `bytes_allocated = new_size`.
But in reality, you have:
* after the call to `Free(old_ptr, old_size)`: `bytes_allocated = 0`
* after the call to `UpdateAllocatedBytes(new_size - old_size)`: `bytes_allocated = new_size - old-size`.

And this is visible in the test results btw: https://travis-ci.org/apache/arrow/jobs/447042467#L1094

xhochy replied:
🤦‍♂️ I was running the release tests while having a debug build. That explains my confusion.

xhochy replied:
Fixed now

pitrou commented on cpp/src/arrow/allocator.h:
```diff
@@ -85,6 +85,41 @@ class stl_allocator {
+
+  Status Reallocate(int64_t old_size, int64_t new_size, uint8_t** ptr) override {
+    uint8_t* old_ptr = *ptr;
+    *ptr = alloc_.allocate(new_size);
```
Also catch `std::bad_alloc` here.

pitrou commented on cpp/src/arrow/allocator-test.cc:
```diff
@@ -23,9 +23,26 @@
+  STLMemoryPool<std::allocator<uint8_t>> pool(allocator);
+
+  uint8_t* data = nullptr;
+  ASSERT_OK(pool.Allocate(100, &data));
```
You should also test `Reallocate` and its effect on `bytes_allocated`.

wesm commented on cpp/src/arrow/allocator.h:
```diff
@@ -85,6 +85,41 @@ class stl_allocator {
+class STLMemoryPool : public MemoryPool {
+ public:
+  explicit STLMemoryPool(const Allocator& alloc) : alloc_(alloc) {}
+  ~STLMemoryPool() override = default;
```
This dtor may not be needed

wesm commented on cpp/src/arrow/memory_pool.h:
```diff
@@ -25,6 +25,35 @@
+  }
+
+ protected:
+  std::atomic<int64_t> bytes_allocated_;
```
Need to include atomic in this header now

xhochy commented:
The test failure is because I have removed the `DCHECK` from the MemoryPoolStats while I moved them to the header. As we don't allow `DCHECK` in the header, I'm currently tending to removing this test.

wesm commented:
I think it's OK to remove that test

pitrou commented:
The "death test" is failing on AppVeyor.

xhochy commented:
AppVeyor partly succeeded here: https://ci.appveyor.com/project/xhochy/arrow/builds/20006144 

Seems like there is some AppVeyor-internal problem currently.
